OscLib/OscCalcNuFast.cxx

```diff
@@ -0,0 +1,295 @@
+    return P(from,to,Eigen::Map<const Eigen::ArrayXd>(E.data(),E.size()));
+  }
+  
+  template<typename T> template<class VT, class KVT>
+  VT _OscCalcNuFast<T>::RecomputeProbabilityMatrix(int from, int to, KVT E) {
+    // Now we need to interface internally used NuFast variables with the OscCalc members.
+    #define s12sq this->fPre.s12sq
+    #define s13sq this->fPre.s13sq
+    #define s23sq this->fPre.s23sq
+    #define sind this->fPre.sind
+    #define cosd this->fPre.cosd
+    #define c12sq this->fPre.c12sq
+    #define c13sq this->fPre.c13sq
+    #define c23sq this->fPre.c23sq
+    
+    #define delta this->fdCP
+    #define Dmsq21 this->fDmsq21
+    #define Dmsq32 this->fDmsq32
+    #define L this->fL
+    #define rho this->fRho
+    #define Ye this->fYe
+    #define N_Newton this->fNNewton
+    #define probs_returned this->fCurrProbs
+    
+    // --------------------------------------------------------------------- //
+    // First calculate useful simple functions of the oscillation parameters //
+    // --------------------------------------------------------------------- //
+    
+    // NOvA doesn't use dmsq31 by convention, but we can compute it if we know dmsq32 and dmsq21.
+    const T
+    Dmsq31 = Dmsq32 + Dmsq21, // dmsq31 = msq3 - msq1 = (msq3-msq2) + (msq2-msq1) = dmsq32 + dmsq21.
+    Dmsqee = Dmsq31 - s12sq * Dmsq21,
+    
+    c13sqxs12sq = c13sq * s12sq,
+    c13sqxs23sq = c13sq * s23sq,
+    c12sqxc23sq = c12sq * c23sq,
+    s13sqxs12sqxs23sq = s13sq * s12sq * s23sq,
+    
+    Jrr = sqrt(c12sqxc23sq * s13sqxs12sqxs23sq),
+    Jmatter_first = 8 * Jrr * c13sq * sind 
+                      * Dmsq21 * Dmsq31 * (Dmsq31 - Dmsq21),
+    
+    Um2sq_first = c12sqxc23sq + s13sqxs12sqxs23sq - 2 * Jrr * cosd,
+    See = Dmsq21+Dmsq31 - Dmsq21 * c13sqxs12sq - Dmsq31 * s13sq,
+    Tee = Dmsq21 * Dmsq31 * (c13sq - c13sqxs12sq);
+    
+    const VT
+    Amatter = (Ye*rho*YerhoE2a) * E,
+    
+    // calculate A, B, C, See, Tee, and part of Tmm
+    C = Amatter * Tee,
+    A = Dmsq21+Dmsq31 + Amatter,
+    
+    // ---------------------------------- //
+    // Get lambda3 from lambda+ of MP/DMP //
+    // ---------------------------------- //
+    xmat = Amatter / Dmsqee;
+    VT lambda3 = Dmsq31 + (0.5 * Dmsqee) * (xmat - 1 + sqrt((1-xmat) * (1-xmat) + (4*s13sq) * xmat));
+    
+    // ---------------------------------------------------------------------------- //
+    // Newton iterations to improve lambda3 arbitrarily, if needed, (B needed here) //
+    // ---------------------------------------------------------------------------- //
+    const VT B = Dmsq21*Dmsq31 + Amatter * See; // B is only needed for N_Newton >= 1
+    for (int i = 0; i < N_Newton; i++) {
+      lambda3 = (lambda3 * lambda3 * (lambda3 + lambda3 - A) + C) / (lambda3 * (2 * (lambda3 - A) + lambda3) + B); // this strange form prefers additions to multiplications
+    }
+    
```

Comment:
If this block is only used sometimes, it should be inside an `if` so that the derivatives aren't computed by Stan when not used

Reply:
I stuck this block inside an `if` and confirmed that the spectra still converge. Actually because I got curious, I just discovered that there is no (well, some extremely small) difference between 0 and 1 Newton iterations by testing spectra. I'm going to set the default to 0 Newton iterations.